File: smoke_tests/scala/deserialization.scala
// Insecure Deserialization vulnerabilities in Scala
package com.example.security

import java.io._
import java.util.Base64

class DeserializationVulnerabilities {

  // Test 1: Java ObjectInputStream
  def deserializeObject(data: Array[Byte]): Any = {
    // VULNERABLE: Deserializing untrusted data
    val bis = new ByteArrayInputStream(data)
    val ois = new ObjectInputStream(bis)
    ois.readObject()
  }

  // Test 2: Base64 encoded serialized object
  def deserializeBase64(encoded: String): Any = {
    val data = Base64.getDecoder.decode(encoded)
    // VULNERABLE: Base64 decoded deserialization
    val bis = new ByteArrayInputStream(data)
    val ois = new ObjectInputStream(bis)
    ois.readObject()
  }

  // Test 3: File-based deserialization
  def loadObjectFromFile(path: String): Any = {
    // VULNERABLE: File could be tampered
    val fis = new FileInputStream(path)
    val ois = new ObjectInputStream(fis)
    ois.readObject()
  }

  // Test 4: Network deserialization
  def receiveObject(socket: java.net.Socket): Any = {
    // VULNERABLE: Network data deserialization
    val ois = new ObjectInputStream(socket.getInputStream)
    ois.readObject()
  }

  // Test 5: Kryo deserialization
  def deserializeKryo(data: Array[Byte]): Any = {
    // VULNERABLE: No class registration in Kryo
    val kryo = new com.esotericsoftware.kryo.Kryo()
    kryo.setRegistrationRequired(false)
    val input = new com.esotericsoftware.kryo.io.Input(data)
    kryo.readClassAndObject(input)
  }

  // Test 6: Play JSON with type info
  def deserializePlayJson(json: String): Any = {
    // VULNERABLE: Type from JSON
    import play.api.libs.json._
    val parsed = Json.parse(json)
    val className = (parsed \ "type").as[String]
    val clazz = Class.forName(className)
    // Deserialize with type
    parsed
  }

  // Test 7: Pickle deserialization
  def deserializePickle(data: Array[Byte]): Any = {
    // VULNERABLE: Scala pickling
    // scala.pickling.Unpickle[Any].fromBytes(data)
    deserializeObject(data)
  }

  // Test 8: Akka serialization
  def deserializeAkka(data: Array[Byte]): Any = {
    // VULNERABLE: Akka serialization from untrusted source
    // system.serialization.deserialize(data, classOf[Any])
    deserializeObject(data)
  }

  // Test 9: Cookie deserialization
  def deserializeCookie(cookieValue: String): Any = {
    val data = Base64.getDecoder.decode(cookieValue)
    // VULNERABLE: Cookie data from client
    val bis = new ByteArrayInputStream(data)
    val ois = new ObjectInputStream(bis)
    ois.readObject()
  }

  // Test 10: Message queue deserialization
  def processMessage(messageBody: Array[Byte]): Any = {
    // VULNERABLE: Message from queue
    val bis = new ByteArrayInputStream(messageBody)
    val ois = new ObjectInputStream(bis)
    ois.readObject()
  }

  // Test 11: Cache deserialization
  def getCachedObject(key: String): Any = {
    val cachedData = getFromCache(key)
    // VULNERABLE: Cache could be poisoned
    val bis = new ByteArrayInputStream(cachedData)
    val ois = new ObjectInputStream(bis)
    ois.readObject()
  }

  // Test 12: circe JSON with runtime types
  def deserializeCirce(json: String, typeName: String): Any = {
    // VULNERABLE: Type name from user
    val clazz = Class.forName(typeName)
    // io.circe.parser.decode(json)(decoder)
    json
  }

  private def getFromCache(key: String): Array[Byte] = Array.empty
}

File: smoke_tests/kotlin/deserialization.kt
// Insecure Deserialization vulnerabilities in Kotlin
package com.example.security

import java.io.*
import java.util.Base64
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.gson.Gson

class DeserializationVulnerabilities {

    private val objectMapper = ObjectMapper()
    private val gson = Gson()

    // Test 1: Java ObjectInputStream
    fun deserializeObject(data: ByteArray): Any? {
        // VULNERABLE: Deserializing untrusted data
        val bis = ByteArrayInputStream(data)
        val ois = ObjectInputStream(bis)
        return ois.readObject()
    }

    // Test 2: Base64 encoded serialized object
    fun deserializeBase64(encoded: String): Any? {
        val data = Base64.getDecoder().decode(encoded)
        // VULNERABLE: Base64 decoded deserialization
        val bis = ByteArrayInputStream(data)
        val ois = ObjectInputStream(bis)
        return ois.readObject()
    }

    // Test 3: File-based deserialization
    fun loadObjectFromFile(path: String): Any? {
        // VULNERABLE: File could be tampered
        val fis = FileInputStream(path)
        val ois = ObjectInputStream(fis)
        return ois.readObject()
    }

    // Test 4: Network deserialization
    fun receiveObject(socket: java.net.Socket): Any? {
        // VULNERABLE: Network data deserialization
        val ois = ObjectInputStream(socket.getInputStream())
        return ois.readObject()
    }

    // Test 5: Jackson with type info
    fun deserializeJson(json: String): Any {
        // VULNERABLE: Default typing enabled
        objectMapper.enableDefaultTyping()
        return objectMapper.readValue(json, Any::class.java)
    }

    // Test 6: Jackson polymorphic
    fun deserializePolymorphic(json: String, typeField: String): Any {
        // VULNERABLE: Type from JSON
        val node = objectMapper.readTree(json)
        val className = node.get(typeField).asText()
        val clazz = Class.forName(className)
        return objectMapper.readValue(json, clazz)
    }

    // Test 7: YAML deserialization
    fun deserializeYaml(yaml: String): Any {
        // VULNERABLE: YAML can deserialize arbitrary objects
        val yamlParser = org.yaml.snakeyaml.Yaml()
        return yamlParser.load(yaml)
    }

    // Test 8: XML deserialization with XStream
    fun deserializeXml(xml: String): Any {
        // VULNERABLE: XStream default allows arbitrary types
        val xstream = com.thoughtworks.xstream.XStream()
        return xstream.fromXML(xml)
    }

    // Test 9: Cookie deserialization
    fun deserializeCookie(cookieValue: String): Any? {
        val data = Base64.getDecoder().decode(cookieValue)
        // VULNERABLE: Cookie data from client
        val bis = ByteArrayInputStream(data)
        val ois = ObjectInputStream(bis)
        return ois.readObject()
    }

    // Test 10: Session deserialization
    fun deserializeSession(sessionData: ByteArray): Map<String, Any>? {
        // VULNERABLE: Session data could be manipulated
        val bis = ByteArrayInputStream(sessionData)
        val ois = ObjectInputStream(bis)
        @Suppress("UNCHECKED_CAST")
        return ois.readObject() as? Map<String, Any>
    }

    // Test 11: Message queue deserialization
    fun processMessage(messageBody: ByteArray): Any? {
        // VULNERABLE: Message from queue
        val bis = ByteArrayInputStream(messageBody)
        val ois = ObjectInputStream(bis)
        return ois.readObject()
    }

    // Test 12: Kryo deserialization
    fun deserializeKryo(data: ByteArray): Any {
        val kryo = com.esotericsoftware.kryo.Kryo()
        // VULNERABLE: No class registration
        kryo.isRegistrationRequired = false
        val input = com.esotericsoftware.kryo.io.Input(data)
        return kryo.readClassAndObject(input)
    }

    // Test 13: Cache deserialization
    fun getCachedObject(key: String): Any? {
        val cachedData = getFromCache(key)
        // VULNERABLE: Cache could be poisoned
        val bis = ByteArrayInputStream(cachedData)
        val ois = ObjectInputStream(bis)
        return ois.readObject()
    }

    private fun getFromCache(key: String): ByteArray = byteArrayOf()
}
